Themis/Models/Repository.swift

```diff
@@ -132,6 +135,53 @@ class Node: Hashable, ObservableObject {
     func hash(into hasher: inout Hasher) {
         hasher.combine(path)
     }
+
+    @MainActor
+    public func calculateDiff(templateParticipationId: Int) async {
+        // only calculate once
+        if diffAdded != nil && diffRemoved != nil {
+            return
+        }
+        guard let code = self.code else {
+            return
+        }
+        if templateCode == nil {
+            await fetchTemplateCode(templateParticipationId: templateParticipationId)
+        }
+        guard let tcode = templateCode else {
+            return
+        }
+        let base = tcode.components(separatedBy: .newlines)
+        let new = code.components(separatedBy: .newlines)
+
+        let diff = new.difference(from: base)
+        var added = [Int]()
+        var removed = [Int]()
+        diff.forEach { change in
+            switch change {
+            case .insert(offset: let offset, element: _, associatedWith: _):
+                added.append(offset + 1) // +1 as lines starting from 1
+            case .remove(offset: let offset, element: _, associatedWith: _):
+                removed.append(offset + 1)
+            }
+        }
+        self.diffAdded = added
+        self.diffRemoved = removed
+        print("added: \(diffAdded ?? [])") // print as long as no highlighting exists
+        print("removed: \(diffRemoved ?? [])")
+    }
+
+    private func fetchTemplateCode(templateParticipationId: Int) async {
+        if templateCode != nil { return } else {
```

Comment:
Use `guard let` at the start for less negation
```suggestion
        guard let templateCode else { return }
```

Reply:
Option so that Swiftlint will also not complain about the unused variable `code`:
```
        guard let _ = self.code else {
            return
        }
```

Reply:
Ok, nevermind, Swiftlint will complain about every possible solution to this